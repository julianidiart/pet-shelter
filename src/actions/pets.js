import database from "../firebase";
import moment from "moment";

// ADD_PET
export const addPet = pet => ({
  type: "ADD_PET",
  pet
});

export const startAddPet = (petData = {}) => {
  return dispatch => {
    const { name = "", birthDate = moment() } = petData;
    const pet = { name, birthDate };
    database
      .ref("pet")
      .push(pet)
      .then(ref => {
        dispatch(addPet({ id: ref.key, ...pet }));
      });
  };
};

// REMOVE_PET
export const removePet = id => ({
  type: "REMOVE_PET",
  id
});

// EDIT_PET
export const editPet = (id, pet) => ({
  type: "EDIT_PET",
  id,
  pet
});
