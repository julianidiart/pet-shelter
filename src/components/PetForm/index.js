import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

export default class PetForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.pet ? props.pet.name : "",
      birthDate: props.pet ? moment(props.pet.birthDate) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onNameChange = e => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };

  onDateChange = birthDate => {
    if (birthDate) {
      this.setState(() => ({ birthDate }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (this.state.name) {
      const error = "";
      this.setState(() => ({ error }));
      this.props.onSubmit({
        name: this.state.name,
        birthDate: this.state.birthDate.valueOf()
      });
    } else {
      const error = "The pet should have a name!";
      this.setState(() => ({ error }));
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            autoFocus
            onChange={this.onNameChange}
            placeholder="Pet name"
            type="text"
            value={this.state.name}
          />
          <SingleDatePicker
            date={this.state.birthDate}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat={() => "DD/MM/YYYY"}
          />
          <button>Save Pet</button>
        </form>
      </div>
    );
  }
}
