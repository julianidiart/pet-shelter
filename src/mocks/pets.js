export default [
  { id: "1", name: "Marla" },
  { id: "2", name: "Bimba" },
  { id: "3", name: "Aimee" },
  { id: "4", name: "Maia" },
  { id: "5", name: "Emma" },
  { id: "6", name: "Tommy" },
  { id: "7", name: "Bello Bruno" },
  { id: "8", name: "Bubu" },
  { id: "9", name: "Blaky" }
];
