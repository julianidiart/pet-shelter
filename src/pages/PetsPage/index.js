import React from "react";

import PetsList from "../../components/PetsList";

export default props => (
  <div>
    <button onClick={() => props.history.push("/add-pet")}>Add pet</button>
    <PetsList />
  </div>
);
