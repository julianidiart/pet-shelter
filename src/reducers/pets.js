const petsReducerDefaultState = [];

export default (state = petsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_PET":
      return [...state, action.pet];
    case "REMOVE_PET":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_PET":
      return state.map(pet => {
        if (pet.id === action.id) {
          return {
            ...pet,
            ...action.pet
          };
        } else {
          return pet;
        }
      });
    case "SET_PETS":
      return action.pets;
    default:
      return state;
  }
};
