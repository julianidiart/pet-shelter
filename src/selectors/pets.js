import moment from "moment";

export default (pets, { text, sortBy, startDate, endDate, filter }) => {
  return pets
    .filter(pet => {
      const birthdateMoment = moment(pet.birthdate);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(birthdateMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(birthdateMoment, "day")
        : true;
      const textMatch =
        pet.name.toLowerCase().includes(text.toLowerCase()) ||
        pet.place.toLowerCase().includes(text.toLowerCase()) ||
        pet.breed.toLowerCase().includes(text.toLowerCase()) ||
        (pet.comments &&
          pet.comments.toLowerCase().includes(text.toLowerCase())) ||
        (pet.color
          ? pet.color.toLowerCase().includes(text.toLowerCase())
          : false) ||
        pet.chip
          .toString()
          .toLowerCase()
          .includes(text.toLowerCase());
      let filterMatch = true;
      const adoptedOrPassedAway = pet.adopted || pet.passedAway;
      switch (filter) {
        case "sterilized":
          filterMatch = !adoptedOrPassedAway && pet.sterilized === true;
          break;
        case "non-sterilized":
          filterMatch = !adoptedOrPassedAway && pet.sterilized === false;
          break;
        case "chipped":
          filterMatch = !adoptedOrPassedAway && pet.chip !== "";
          break;
        case "non-chipped":
          filterMatch = !adoptedOrPassedAway && pet.chip === "";
          break;
        case "adopted":
          filterMatch = pet.adopted === true;
          break;
        case "passed-away":
          filterMatch = pet.passedAway === true;
          break;
        default:
          filterMatch = !adoptedOrPassedAway;
      }

      return startDateMatch && endDateMatch && textMatch && filterMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.birthdate < b.birthdate ? 1 : -1;
      }
      return a.name < b.name ? 1 : -1;
    });
};
