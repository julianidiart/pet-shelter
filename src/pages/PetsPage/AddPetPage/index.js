import React, { Component } from "react";
import { connect } from "react-redux";
import PetForm from "../../../components/PetForm";
import { startAddPet } from "../../../actions/pets";

export class AddPetPage extends Component {
  onSubmit = pet => {
    this.props.startAddPet(pet);
    this.props.history.push("/pets");
  };
  render() {
    return (
      <div>
        <PetForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPet: pet => dispatch(startAddPet(pet))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddPetPage);
