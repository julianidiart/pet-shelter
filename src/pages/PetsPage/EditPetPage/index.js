import React, { Component } from "react";
import { connect } from "react-redux";
import PetForm from "../../../components/PetForm";
import { editPet, removePet } from "../../../actions/pets";

export class EditPetPage extends Component {
  onSubmit = pet => {
    this.props.editPet(this.props.pet.id, pet);
    this.props.history.push("/pets");
  };
  onClick = () => {
    this.props.removePet(this.props.pet.id);
    this.props.history.push("/pets");
  };
  render() {
    return (
      <div>
        <PetForm pet={this.props.pet} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    pet: state.pets.find(pet => pet.id === props.match.params.id)
  };
};

const mapDispatchToProps = dispatch => ({
  editPet: (id, pet) => dispatch(editPet(id, pet)),
  removePet: id => dispatch(removePet(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPetPage);
