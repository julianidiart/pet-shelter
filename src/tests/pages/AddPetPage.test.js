import React from "react";
import { shallow } from "enzyme";
import { AddPetPage } from "../../pages/PetsPage/AddPetPage";
import PetForm from "../../components/PetForm";
import pets from "../fixtures/pets";

let startAddPet, history, wrapper;

beforeEach(() => {
  startAddPet = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddPetPage startAddPet={startAddPet} history={history} />);
});

test("should render AddPetPage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
  wrapper.find(PetForm).prop("onSubmit")(pets[0]);
  expect(history.push).toHaveBeenLastCalledWith("/pets");
  expect(startAddPet).toHaveBeenLastCalledWith(pets[0]);
});
