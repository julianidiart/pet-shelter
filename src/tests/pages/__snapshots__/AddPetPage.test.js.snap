// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddPetPage correctly 1`] = `
<div>
  <PetForm
    onSubmit={[Function]}
  />
</div>
`;
